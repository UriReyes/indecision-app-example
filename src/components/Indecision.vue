<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input v-model="question" type="text" placeholder="Hazme una pregunta" />
    <p style="margin-top: 10px">
      Recuerda terminar con signo de interrogación (?)
    </p>
    <div v-if="isValidQuestion">
      <h2>{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        const response = await fetch("https://yesno.wtf/api");
        const { answer, image } = await response.json();
        this.answer = this.translateAnswer(answer);
        this.img = image;
      } catch (error) {
        console.log("IndecisionApp: " + error);
        this.answer = "No se pudo cargar el API";
        this.img = null;
      }
    },
    translateAnswer(answer) {
      switch (answer) {
        case "yes":
          return "Sí";
        case "no":
          return "No!";
        case "maybe":
          return "Tal Vez";
      }
    },
    backToDefault() {
      this.isValidQuestion = false;
      this.answer = "Pensando...";
      this.img = null;
    },
  },
  watch: {
    question(value, oldValue) {
      // const regex = new RegExp("^[¿]([a-zA-Z\s]|[à-ú\s]|[À-Ú\s]){2,}[?]");
      // console.log(regex.test(value));
      console.log(value);
      this.backToDefault();
      if (!value.includes("?")) return;
      //TODO: Realizar petición http
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
