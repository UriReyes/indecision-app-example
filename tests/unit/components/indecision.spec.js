import { shallowMount } from '@vue/test-utils'
import Indecision from "@/components/Indecision.vue"

describe('Indecision Component', () => {
    let wrapper
    let clgSpy
    global.fetch = jest.fn(() => Promise.resolve({
        json: () => Promise.resolve({
            "answer": "yes",
            "forced": false,
            "image": "https://yesno.wtf/assets/yes/2.gif"
        })
    }))
    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console, 'log')
        jest.clearAllMocks();
    })

    test('should match with the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('write in the input don´t must trigger nothing (console.log)', async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Hola mundo')
        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).not.toHaveBeenCalled()
    })

    test('write the symbol "?" in the input must trigger the getAnswer method', async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')
        const input = wrapper.find('input')
        await input.setValue('Seré millonario?')
        expect(clgSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalled()
    })

    test('testing the getAnswer method', async () => {
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeTruthy()
        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
        expect(wrapper.vm.answer).toBe('Sí')
    })
    test('testing the getAnswer method - Simulate API fail', async () => {
        //TODO: fallo en el API
        fetch.mockImplementationOnce(() => Promise.reject('API is down'))
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img')
        expect(img.exists()).toBeFalsy()
        expect(wrapper.vm.answer).toBe('No se pudo cargar el API')
    })
})