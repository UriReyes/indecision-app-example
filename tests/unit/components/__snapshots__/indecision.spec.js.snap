// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component should match with the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p style="margin-top: 10px;"> Recuerda terminar con signo de interrogaci√≥n (?) </p>
  <!--v-if-->
</div>
`;
